### PR TITLE
Extend the infrastructure provider with `StartNodes` and `StopTestnet`

### PR DESCRIPTION
**Pull request by sergio-mena**
Contributes to #55

This PR is reusing tendermint/tendermint#9801 and driving it to completion:
* Two new methods added to infrastructure provider with `StartNodes` and `StopTestnet`
* Refactored the docker provider and the runner logic to use these two interface methods
* Prepared a new provider, `digitalocean` to implement these methods in a subsequent PR
* Distingush between internal and external IP
  * on DO, they will be the same
  * on Docker, the external IP will be `localhost` for all nodes (the proxy address)

---

#### PR checklist

- [ ] Tests written/updated
- [ ] Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [ ] Updated relevant documentation (`docs/` or `spec/`) and code comments

### CODE REVIEW

File: test/e2e/runner/start.go [hunk 1]
```diff
@@ -102,15 +119,16 @@ func Start(ctx context.Context, testnet *e2e.Testnet) error {
 
 		logger.Info("Starting catch up node", "node", node.Name, "height", node.StartAt)
 
-		if err := execCompose(testnet.Dir, "up", "-d", node.Name); err != nil {
+		err := p.StartNodes(context.Background(), node)
```
Comment (lasarojc) [inline test/e2e/runner/start.go:122]: in the previous loop, logically, we should first `waitForHeight(... node.StartAt ...)` and then update `networkHeight`.

File: test/e2e/runner/main.go [hunk 2]
```diff
@@ -186,7 +196,7 @@ func NewCLI() *CLI {
 			if err != nil {
 				return err
 			}
-			return Start(cmd.Context(), cli.testnet)
+			return Start(cmd.Context(), cli.testnet, cli.infp)
```
Comment (lasarojc) [inline test/e2e/runner/main.go:199]: in the short description, above, update to include DO?

File: test/e2e/runner/main.go [hunk 3]
```diff
@@ -211,7 +221,7 @@ func NewCLI() *CLI {
 		Short: "Stops the Docker testnet",
 		RunE: func(cmd *cobra.Command, args []string) error {
 			logger.Info("Stopping testnet")
-			return execCompose(cli.testnet.Dir, "down")
+			return cli.infp.StopTestnet(context.Background())
```
Comment (lasarojc) [inline test/e2e/runner/main.go:224]: in the short description above, update to include DO?

File: test/e2e/runner/main.go [hunk 4]
```diff
@@ -266,15 +276,15 @@ func NewCLI() *CLI {
 		Use:   "logs",
 		Short: "Shows the testnet logs",
 		RunE: func(cmd *cobra.Command, args []string) error {
-			return execComposeVerbose(cli.testnet.Dir, "logs")
+			return docker.ExecComposeVerbose(context.Background(), cli.testnet.Dir, "logs")
```
Comment (lasarojc) [inline test/e2e/runner/main.go:279]: will this eventually also work for DO?
maybe add a TODO?

File: test/e2e/runner/main.go [hunk 5]
```diff
@@ -266,15 +276,15 @@ func NewCLI() *CLI {
...
 		},
 	})
 
 	cli.root.AddCommand(&cobra.Command{
 		Use:   "tail",
 		Short: "Tails the testnet logs",
 		RunE: func(cmd *cobra.Command, args []string) error {
-			return execComposeVerbose(cli.testnet.Dir, "logs", "--follow")
+			return docker.ExecComposeVerbose(context.Background(), cli.testnet.Dir, "logs", "--follow")
```
Comment (lasarojc) [inline test/e2e/runner/main.go:287]: ditto

File: test/e2e/runner/perturb.go [hunk 6]
```diff
@@ -28,7 +29,7 @@ func Perturb(ctx context.Context, testnet *e2e.Testnet) error {
 // after recovering.
 func PerturbNode(ctx context.Context, node *e2e.Node, perturbation e2e.Perturbation) (*rpctypes.ResultStatus, error) {
 	testnet := node.Testnet
-	out, err := execComposeOutput(testnet.Dir, "ps", "-q", node.Name)
+	out, err := docker.ExecComposeOutput(context.Background(), testnet.Dir, "ps", "-q", node.Name)
```
Comment (lasarojc) [inline test/e2e/runner/perturb.go:32]: will this eventually work also for DO?

File: test/e2e/pkg/exec/exec.go [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+}
+
+// CommandOutput executes a shell command and return the command's output.
+func CommandOutput(ctx context.Context, args ...string) ([]byte, error) {
+	//nolint: gosec
+	// G204: Subprocess launched with a potential tainted input or cmd arguments
+	cmd := osexec.CommandContext(ctx, args[0], args[1:]...)
+	out, err := cmd.CombinedOutput()
+	switch err := err.(type) {
+	case nil:
```
Comment (lasarojc) [inline test/e2e/pkg/exec/exec.go:23]: is it just me or do you also find it weird that you can test a type against nil?

File: test/e2e/pkg/exec/exec.go [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+	osexec "os/exec"
+)
+
+// Command executes a shell command.
+func Command(ctx context.Context, args ...string) error {
+	_, err := CommandOutput(ctx, args...)
+	return err
+}
+
+// CommandOutput executes a shell command and return the command's output.
```
Comment (lasarojc) [inline test/e2e/pkg/exec/exec.go:16]:
```suggestion
// CommandOutput executes a shell command and returns the command's output.
```

**Review by lasarojc — APPROVED** [review]
Thanks for driving this. It will be a huge improvement for our QA.

Made a few suggestions, but nothing blocking.

File: test/e2e/runner/perturb.go (hunk 6 above)
Comment (thanethomson) [reply test/e2e/runner/perturb.go:32]: I think the point of the PR was just to start putting in some of the initial plumbing to get the DO nodes to start/stop. The original intention was to expand this functionality with perturbations in follow-up PRs.

**Review by thanethomson — APPROVED** [review]

File: test/e2e/pkg/exec/exec.go (hunk 7 above)
Comment (sergio-mena) [reply test/e2e/pkg/exec/exec.go:23]: This was coming from William's patch, so I assumed it was good.
I've just taken a look at our code base, and we do this quite often, e.g. [here,](https://github.com/cometbft/cometbft/blob/bd9d7e/light/client.go#L731) [here](https://github.com/cometbft/cometbft/blob/bd9d7e/evidence/reactor.go#L88), or [here](https://github.com/cometbft/cometbft/blob/4790ea3e46475064d5475c787427ae926c5a9e94/light/detector.go#L59).

Comment (sergio-mena) [reply test/e2e/pkg/exec/exec.go:23]: Seems to be a standard way to handle errors...

Comment (lasarojc) [reply test/e2e/pkg/exec/exec.go:23]: It seems to be one of Go's idiosyncrasies

File: test/e2e/runner/main.go [hunk 9]
```diff
@@ -30,6 +31,7 @@ type CLI struct {
 	testnet  *e2e.Testnet
 	preserve bool
 	infp     infra.Provider
+	ifd      e2e.InfrastructureData
```
[the remark this answers is not in the document]
Comment (sergio-mena) [reply test/e2e/runner/main.go:34]: Good point... let me see what I can do (I need to check if that's OK with the UTs below)

[the remark this answers is not in the document]
Comment (sergio-mena) [reply test/e2e/runner/main.go]: I was a bit more involved than expected, but the result is worth it, I think

File: test/e2e/runner/main.go (hunk 2 above)
Comment (sergio-mena) [reply test/e2e/runner/main.go:204]: Good catch. To avoid having to come back to it at every subsequent PR, I just removed references to Docker

File: test/e2e/runner/main.go (hunk 3 above)
Comment (sergio-mena) [reply test/e2e/runner/main.go:229]: Removed reference to Docker

File: test/e2e/runner/perturb.go (hunk 6 above)
Comment (sergio-mena) [reply test/e2e/runner/perturb.go:32]: To add on to @thanethomson's comment, `PerturbNode` will eventually become a method in `infra.Provider`, whose Docker version will be + or - it current implementation; and we will have to come up with mechanisms for implementing it for a DO testnet.

File: test/e2e/runner/start.go (hunk 1 above)
Comment (sergio-mena) [reply test/e2e/runner/start.go:122]: Do you mean the loop starting at line 54? Those nodes start all at height zero.
I had to change the logic a bit here, to make the DO way of doing things compatible with Docker:
* Before: nodes that start at height zero were started one by one. That is very inefficient with a large testnet (think of the 200-node testnet)
* After: nodes starting at height zero are all started at the same time. Then we wait for all of them to be started (`waitForNode` with height zero)
